packages/apollo-voyager-metrics/src/index.ts

```diff
@@ -33,64 +31,138 @@ const serverResponseMetric = new Prometheus.Histogram({
   labelNames: ['request_type', 'error']
 })
 
-export function getMetrics (req: IncomingMessage, res: Response) {
-  res.set('Content-Type', Prometheus.register.contentType)
-  res.end(Prometheus.register.metrics())
+interface ResolverObject {
+  [key: string]: IFieldResolver<any, any>
+}
 
-  resolverTimingMetric.reset()
-  resolverRequestsMetric.reset()
-  serverResponseMetric.reset()
+export function enableDefaultMetricsColleciton () {
+  if (!promMetricsEnabled) {
+    Prometheus.collectDefaultMetrics()
+    promMetricsEnabled = true
+  }
 }
 
-export function responseLoggingMetric (req: IncomingMessage, res: ResponseWithVoyagerMetrics, next: NextFunction) {
-  const requestMethod = req.method as string
+export function wrapResolversForMetrics (resolvers: {[key: string]: ResolverObject}): {[key: string]: ResolverObject} {
+  const output: {[key: string]: ResolverObject} = {}
+  for (const typeKey in resolvers) {
+    if (!resolvers.hasOwnProperty(typeKey)) {
+      continue
+    }
+    const fieldResolversForType = resolvers[typeKey]
+    output[typeKey] = {}
+    for (const fieldKey in fieldResolversForType) {
+      if (!fieldResolversForType.hasOwnProperty(fieldKey)) {
+        continue
+      }
+      const resolverForField = fieldResolversForType[fieldKey]
+      output[typeKey][fieldKey] = wrapSingleResolverForMetrics(resolverForField)
+    }
+  }
 
-  res.requestStart = Date.now()
+  return output
+}
 
-  res.on('finish', onResFinished)
-  res.on('error', onResFinished)
+export function applyResponseLoggingMetricsMiddleware (app: Application) {
+  app.use(responseLoggingMetric as (req: IncomingMessage, res: Response, next: NextFunction) => void)
+}
 
-  if (next) { next() }
+export function applyMetricsMiddleware (app: Application) {
+  app.get('/metrics', getMetrics)
```

Comment:
Same as above. +1  for the idea but I feel like the end user should be able to configure the endpoint. Is there any strong reason why they shouldn't be able to, in terms of APBs or things like that?

Reply:
I did it that way because I saw `applyMiddleware` in other places. Wanted to keep it consistent. No strong opinion though. Can do it that way.

Reply:
Sounds good. Just for context, the `applyMiddleware` pattern comes from the various apollo server integrations such as `apollo-server-express`, `apollo-server-hapi` `apollo-server-koa`. <- those libs are actually very thin wrappers that extend the base `ApolloServer` class with a framework specific `applyMiddleware` function that knows how to do the correct setup for a given framework.

We're not passing in any options into `server.applyMiddleware()` in the example apps but it does accept an options object with some things like a path. We can come back to it later but definitely we'll want that at some stage. We may eventually want to go down a similar route where we can support more than just express. 🙂